File: app/src/main/java/com/kunzisoft/keepass/dialogs/GroupEditDialogFragment.java
/*
 * Copyright 2017 Brian Pellin, Jeremy Jamet / Kunzisoft.
 *     
 * This file is part of KeePass DX.
 *
 *  KeePass DX is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  KeePass DX is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with KeePass DX.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.kunzisoft.keepass.dialogs;

import android.app.Dialog;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.support.v7.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.kunzisoft.keepass.R;
import com.kunzisoft.keepass.app.App;
import com.kunzisoft.keepass.database.Database;
import com.kunzisoft.keepass.database.PwGroup;
import com.kunzisoft.keepass.database.PwIcon;

import static com.kunzisoft.keepass.dialogs.GroupEditDialogFragment.EditGroupDialogAction.CREATION;
import static com.kunzisoft.keepass.dialogs.GroupEditDialogFragment.EditGroupDialogAction.UPDATE;
import static com.kunzisoft.keepass.dialogs.GroupEditDialogFragment.EditGroupDialogAction.getActionFromOrdinal;

public class GroupEditDialogFragment extends DialogFragment
        implements IconPickerDialogFragment.IconPickerListener {

    public static final String TAG_CREATE_GROUP = "TAG_CREATE_GROUP";

	public static final String KEY_NAME = "KEY_NAME";
	public static final String KEY_ICON = "KEY_ICON";
	public static final String KEY_ACTION_ID = "KEY_ACTION_ID";

	private Database database;

	private EditGroupListener editGroupListener;

    private EditGroupDialogAction editGroupDialogAction;
    private String nameGroup;
    private PwIcon iconGroup;

    private ImageView iconButton;
    private int iconColor;

    public enum EditGroupDialogAction {
        CREATION, UPDATE, NONE;

        public static EditGroupDialogAction getActionFromOrdinal(int ordinal) {
            return EditGroupDialogAction.values()[ordinal];
        }
    }

    public static GroupEditDialogFragment build() {
        Bundle bundle = new Bundle();
        bundle.putInt(KEY_ACTION_ID, CREATION.ordinal());
        GroupEditDialogFragment fragment = new GroupEditDialogFragment();
        fragment.setArguments(bundle);
        return fragment;
    }

    public static GroupEditDialogFragment build(PwGroup group) {
        Bundle bundle = new Bundle();
        bundle.putString(KEY_NAME, group.getName());
        bundle.putParcelable(KEY_ICON, group.getIcon());
        bundle.putInt(KEY_ACTION_ID, UPDATE.ordinal());
        GroupEditDialogFragment fragment = new GroupEditDialogFragment();
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        // Verify that the host activity implements the callback interface
        try {
            // Instantiate the NoticeDialogListener so we can send events to the host
            editGroupListener = (EditGroupListener) context;
        } catch (ClassCastException e) {
            // The activity doesn't implement the interface, throw exception
            throw new ClassCastException(context.toString()
                    + " must implement " + GroupEditDialogFragment.class.getName());
        }
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        assert getActivity() != null;
        LayoutInflater inflater = getActivity().getLayoutInflater();
        View root = inflater.inflate(R.layout.group_edit, null);
        TextView nameField = root.findViewById(R.id.group_name);
        iconButton = root.findViewById(R.id.icon_button);

        // Retrieve the textColor to tint the icon
        int[] attrs = {android.R.attr.textColorPrimary};
        TypedArray ta = getActivity().getTheme().obtainStyledAttributes(attrs);
        iconColor = ta.getColor(0, Color.WHITE);

        // Init elements
        database = App.getDB();
        editGroupDialogAction = EditGroupDialogAction.NONE;
        nameGroup = "";
        iconGroup = database.getPwDatabase().getIconFactory().getFolderIcon();

        if (savedInstanceState != null
                && savedInstanceState.containsKey(KEY_ACTION_ID)
                && savedInstanceState.containsKey(KEY_NAME)
                && savedInstanceState.containsKey(KEY_ICON)) {
            editGroupDialogAction = getActionFromOrdinal(savedInstanceState.getInt(KEY_ACTION_ID));
            nameGroup = savedInstanceState.getString(KEY_NAME);
            iconGroup = savedInstanceState.getParcelable(KEY_ICON);

        } else {

            if (getArguments() != null
                    && getArguments().containsKey(KEY_ACTION_ID))
                editGroupDialogAction = EditGroupDialogAction.getActionFromOrdinal(getArguments().getInt(KEY_ACTION_ID));

            if (getArguments() != null
                    && getArguments().containsKey(KEY_NAME)
                    && getArguments().containsKey(KEY_ICON)) {
                nameGroup = getArguments().getString(KEY_NAME);
                iconGroup = getArguments().getParcelable(KEY_ICON);
            }
        }

        // populate the name
        nameField.setText(nameGroup);
        // populate the icon
        assignIconView();

        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setView(root)
                .setPositiveButton(android.R.string.ok, (dialog, id) -> {
                    String name = nameField.getText().toString();
                    if ( name.length() > 0 ) {
                        editGroupListener.approveEditGroup(
                                editGroupDialogAction,
                                name,
                                iconGroup);

                        GroupEditDialogFragment.this.getDialog().cancel();
                    }
                    else {
                        Toast.makeText(getContext(), R.string.error_no_name, Toast.LENGTH_LONG).show();
                    }
                })
                .setNegativeButton(R.string.cancel, (dialog, id) -> {
                    String name = nameField.getText().toString();
                    editGroupListener.cancelEditGroup(
                            editGroupDialogAction,
                            name,
                            iconGroup);

                    GroupEditDialogFragment.this.getDialog().cancel();
                });

        iconButton.setOnClickListener(v -> {
            IconPickerDialogFragment iconPickerDialogFragment = new IconPickerDialogFragment();
            if (getFragmentManager() != null)
                iconPickerDialogFragment.show(getFragmentManager(), "IconPickerDialogFragment");
        });

        return builder.create();
    }

    private void assignIconView() {
        database.getDrawFactory()
                .assignDatabaseIconTo(
                        getContext(),
                        iconButton,
                        iconGroup,
                        iconColor);
    }

    @Override
    public void iconPicked(Bundle bundle) {
        iconGroup = bundle.getParcelable(IconPickerDialogFragment.KEY_ICON_STANDARD);
        assignIconView();
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        outState.putInt(KEY_ACTION_ID, editGroupDialogAction.ordinal());
        outState.putString(KEY_NAME, nameGroup);
        outState.putParcelable(KEY_ICON, iconGroup);
        super.onSaveInstanceState(outState);
    }

    public interface EditGroupListener {
        void approveEditGroup(EditGroupDialogAction action, String name, PwIcon selectedIcon);
        void cancelEditGroup(EditGroupDialogAction action, String name, PwIcon selectedIcon);
    }
}

File: app/src/main/java/com/kunzisoft/keepass/activities/dialogs/GroupEditDialogFragment.kt
/*
 * Copyright 2019 Jeremy Jamet / Kunzisoft.
 *     
 * This file is part of KeePassDX.
 *
 *  KeePassDX is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  KeePassDX is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with KeePassDX.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.kunzisoft.keepass.activities.dialogs

import android.app.Dialog
import android.graphics.Color
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.activityViewModels
import com.google.android.material.textfield.TextInputLayout
import com.kunzisoft.keepass.R
import com.kunzisoft.keepass.activities.dialogs.GroupEditDialogFragment.EditGroupDialogAction.CREATION
import com.kunzisoft.keepass.activities.dialogs.GroupEditDialogFragment.EditGroupDialogAction.NONE
import com.kunzisoft.keepass.activities.dialogs.GroupEditDialogFragment.EditGroupDialogAction.UPDATE
import com.kunzisoft.keepass.adapters.TagsProposalAdapter
import com.kunzisoft.keepass.database.ContextualDatabase
import com.kunzisoft.keepass.database.element.DateInstant
import com.kunzisoft.keepass.database.element.icon.IconImage
import com.kunzisoft.keepass.model.GroupInfo
import com.kunzisoft.keepass.utils.getParcelableCompat
import com.kunzisoft.keepass.view.DateTimeEditFieldView
import com.kunzisoft.keepass.view.InheritedCompletionView
import com.kunzisoft.keepass.view.TagsCompletionView
import com.kunzisoft.keepass.viewmodels.GroupEditViewModel
import com.tokenautocomplete.FilteredArrayAdapter
import org.joda.time.DateTime

class GroupEditDialogFragment : DatabaseDialogFragment() {

    private val mGroupEditViewModel: GroupEditViewModel by activityViewModels()

    private var mPopulateIconMethod: ((ImageView, IconImage) -> Unit)? = null
    private var mEditGroupDialogAction = NONE
    private var mGroupInfo = GroupInfo()
    private var mGroupNamesNotAllowed: List<String>? = null

    private lateinit var iconButtonView: ImageView
    private var mIconColor: Int = 0
    private lateinit var nameTextLayoutView: TextInputLayout
    private lateinit var nameTextView: TextView
    private lateinit var notesTextLayoutView: TextInputLayout
    private lateinit var notesTextView: TextView
    private lateinit var expirationView: DateTimeEditFieldView
    private lateinit var searchableContainerView: TextInputLayout
    private lateinit var searchableView: InheritedCompletionView
    private lateinit var autoTypeContainerView: ViewGroup
    private lateinit var autoTypeInheritedView: InheritedCompletionView
    private lateinit var autoTypeSequenceView: TextView
    private lateinit var tagsContainerView: TextInputLayout
    private lateinit var tagsCompletionView: TagsCompletionView
    private var tagsAdapter: FilteredArrayAdapter<String>? = null

    enum class EditGroupDialogAction {
        CREATION, UPDATE, NONE;

        companion object {
            fun getActionFromOrdinal(ordinal: Int): EditGroupDialogAction {
                return values()[ordinal]
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        mGroupEditViewModel.onIconSelected.observe(this) { iconImage ->
            mGroupInfo.icon = iconImage
            mPopulateIconMethod?.invoke(iconButtonView, mGroupInfo.icon)
        }

        mGroupEditViewModel.onDateSelected.observe(this) { dateMilliseconds ->
            // Save the date
            mGroupInfo.expiryTime = DateInstant(
                DateTime(mGroupInfo.expiryTime.date)
                    .withMillis(dateMilliseconds)
                    .toDate())
            expirationView.dateTime = mGroupInfo.expiryTime
            if (expirationView.dateTime.type == DateInstant.Type.DATE_TIME) {
                val instantTime = DateInstant(mGroupInfo.expiryTime.date, DateInstant.Type.TIME)
                // Trick to recall selection with time
                mGroupEditViewModel.requestDateTimeSelection(instantTime)
            }
        }

        mGroupEditViewModel.onTimeSelected.observe(this) { viewModelTime ->
            // Save the time
            mGroupInfo.expiryTime = DateInstant(
                DateTime(mGroupInfo.expiryTime.date)
                    .withHourOfDay(viewModelTime.hours)
                    .withMinuteOfHour(viewModelTime.minutes)
                    .toDate(), mGroupInfo.expiryTime.type)
            expirationView.dateTime = mGroupInfo.expiryTime
        }

        mGroupEditViewModel.groupNamesNotAllowed.observe(this) { namesNotAllowed ->
            this.mGroupNamesNotAllowed = namesNotAllowed
        }
    }

    override fun onDatabaseRetrieved(database: ContextualDatabase?) {
        super.onDatabaseRetrieved(database)

        mPopulateIconMethod = { imageView, icon ->
            database?.iconDrawableFactory?.assignDatabaseIcon(imageView, icon, mIconColor)
        }
        mPopulateIconMethod?.invoke(iconButtonView, mGroupInfo.icon)

        searchableContainerView.visibility = if (database?.allowCustomSearchableGroup() == true) {
            View.VISIBLE
        } else {
            View.GONE
        }

        if (database?.allowAutoType() == true) {
            autoTypeContainerView.visibility = View.VISIBLE
        } else {
            autoTypeContainerView.visibility = View.GONE
        }

        tagsAdapter = TagsProposalAdapter(requireContext(), database?.tagPool)
        tagsCompletionView.apply {
            threshold = 1
            setAdapter(tagsAdapter)
        }
        tagsContainerView.visibility = if (database?.allowTags() == true) View.VISIBLE else View.GONE
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        activity?.let { activity ->
            val root = activity.layoutInflater.inflate(R.layout.fragment_group_edit, null)
            iconButtonView = root.findViewById(R.id.group_edit_icon_button)
            nameTextLayoutView = root.findViewById(R.id.group_edit_name_container)
            nameTextView = root.findViewById(R.id.group_edit_name)
            notesTextLayoutView = root.findViewById(R.id.group_edit_note_container)
            notesTextView = root.findViewById(R.id.group_edit_note)
            expirationView = root.findViewById(R.id.group_edit_expiration)
            searchableContainerView = root.findViewById(R.id.group_edit_searchable_container)
            searchableView = root.findViewById(R.id.group_edit_searchable)
            autoTypeContainerView = root.findViewById(R.id.group_edit_auto_type_container)
            autoTypeInheritedView = root.findViewById(R.id.group_edit_auto_type_inherited)
            autoTypeSequenceView = root.findViewById(R.id.group_edit_auto_type_sequence)
            tagsContainerView = root.findViewById(R.id.group_tags_label)
            tagsCompletionView = root.findViewById(R.id.group_tags_completion_view)

            // Retrieve the textColor to tint the icon
            val ta = activity.theme.obtainStyledAttributes(intArrayOf(android.R.attr.textColor))
            mIconColor = ta.getColor(0, Color.WHITE)
            ta.recycle()

            if (savedInstanceState != null
                    && savedInstanceState.containsKey(KEY_ACTION_ID)
                    && savedInstanceState.containsKey(KEY_GROUP_INFO)) {
                mEditGroupDialogAction = EditGroupDialogAction.getActionFromOrdinal(savedInstanceState.getInt(KEY_ACTION_ID))
                mGroupInfo = savedInstanceState.getParcelableCompat(KEY_GROUP_INFO) ?: mGroupInfo
            } else {
                arguments?.apply {
                    if (containsKey(KEY_ACTION_ID))
                        mEditGroupDialogAction = EditGroupDialogAction.getActionFromOrdinal(getInt(KEY_ACTION_ID))
                    if (containsKey(KEY_GROUP_INFO)) {
                        mGroupInfo = getParcelableCompat(KEY_GROUP_INFO) ?: mGroupInfo
                    }
                }
            }

            // populate info in views
            populateInfoToViews(mGroupInfo)

            iconButtonView.setOnClickListener { _ ->
                mGroupEditViewModel.requestIconSelection(mGroupInfo.icon)
            }
            expirationView.setOnDateClickListener = { dateInstant ->
                mGroupEditViewModel.requestDateTimeSelection(dateInstant)
            }

            val builder = AlertDialog.Builder(activity)
            builder.setView(root)
                    .setPositiveButton(android.R.string.ok, null)
                    .setNegativeButton(android.R.string.cancel) { _, _ ->
                        // Do nothing
                    }

            return builder.create()
        }
        return super.onCreateDialog(savedInstanceState)
    }

    override fun onResume() {
        super.onResume()

        // To prevent auto dismiss
        val alertDialog = dialog as AlertDialog?
        if (alertDialog != null) {
            val positiveButton = alertDialog.getButton(Dialog.BUTTON_POSITIVE) as Button
            positiveButton.setOnClickListener {
                retrieveGroupInfoFromViews()
                if (isValid()) {
                    when (mEditGroupDialogAction) {
                        CREATION ->
                            mGroupEditViewModel.approveGroupCreation(mGroupInfo)
                        UPDATE ->
                            mGroupEditViewModel.approveGroupUpdate(mGroupInfo)
                        NONE -> {}
                    }
                    alertDialog.dismiss()
                }
            }
        }
    }

    private fun populateInfoToViews(groupInfo: GroupInfo) {
        mGroupEditViewModel.selectIcon(groupInfo.icon)
        nameTextView.text = groupInfo.title
        notesTextLayoutView.visibility = if (groupInfo.notes == null) View.GONE else View.VISIBLE
        groupInfo.notes?.let {
            notesTextView.text = it
        }
        expirationView.activation = groupInfo.expires
        expirationView.dateTime = groupInfo.expiryTime

        // Set searchable
        searchableView.setValue(groupInfo.searchable)
        // Set auto-type
        autoTypeInheritedView.setValue(groupInfo.enableAutoType)
        autoTypeSequenceView.text = groupInfo.defaultAutoTypeSequence
        // Set Tags
        groupInfo.tags.let { tags ->
            tagsCompletionView.setText("")
            for (i in 0 until tags.size()) {
                tagsCompletionView.addObjectSync(tags.get(i))
            }
        }
    }

    private fun retrieveGroupInfoFromViews() {
        mGroupInfo.title = nameTextView.text.toString()
        // Only if there
        val newNotes = notesTextView.text.toString()
        if (newNotes.isNotEmpty()) {
            mGroupInfo.notes = newNotes
        }
        mGroupInfo.expires = expirationView.activation
        mGroupInfo.expiryTime = expirationView.dateTime
        mGroupInfo.searchable = searchableView.getValue()
        mGroupInfo.enableAutoType = autoTypeInheritedView.getValue()
        mGroupInfo.defaultAutoTypeSequence = autoTypeSequenceView.text.toString()
        mGroupInfo.tags = tagsCompletionView.getTags()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        retrieveGroupInfoFromViews()
        outState.putInt(KEY_ACTION_ID, mEditGroupDialogAction.ordinal)
        outState.putParcelable(KEY_GROUP_INFO, mGroupInfo)
        super.onSaveInstanceState(outState)
    }

    private fun isValid(): Boolean {
        val name = nameTextView.text.toString()
        val error = when {
            name.isEmpty() -> {
                Error(true, R.string.error_no_name)
            }
            mGroupNamesNotAllowed == null -> {
                Error(true, R.string.error_word_reserved)
            }
            mGroupNamesNotAllowed?.find { it.equals(name, ignoreCase = true) } != null -> {
                Error(true, R.string.error_word_reserved)
            }
            else -> {
                Error(false, null)
            }
        }
        error.messageId?.let { messageId ->
            nameTextLayoutView.error = getString(messageId)
        } ?: kotlin.run {
            nameTextLayoutView.error = null
        }
        return !error.isError
    }

    data class Error(val isError: Boolean, val messageId: Int?)

    companion object {

        const val TAG_CREATE_GROUP = "TAG_CREATE_GROUP"
        private const val KEY_ACTION_ID = "KEY_ACTION_ID"
        private const val KEY_GROUP_INFO = "KEY_GROUP_INFO"

        fun create(groupInfo: GroupInfo): GroupEditDialogFragment {
            val bundle = Bundle()
            bundle.putInt(KEY_ACTION_ID, CREATION.ordinal)
            bundle.putParcelable(KEY_GROUP_INFO, groupInfo)
            val fragment = GroupEditDialogFragment()
            fragment.arguments = bundle
            return fragment
        }

        fun update(groupInfo: GroupInfo): GroupEditDialogFragment {
            val bundle = Bundle()
            bundle.putInt(KEY_ACTION_ID, UPDATE.ordinal)
            bundle.putParcelable(KEY_GROUP_INFO, groupInfo)
            val fragment = GroupEditDialogFragment()
            fragment.arguments = bundle
            return fragment
        }
    }
}
